File: android/ac-core/src/main/kotlin/com/microsoft/adaptivecards/core/FeatureFlags.kt
// Copyright (c) Microsoft Corporation. All rights reserved.
// Author: Vikrant Singh (github.com/VikrantSingh01)
// Licensed under the MIT License.

package com.microsoft.adaptivecards.core

/**
 * Feature flags for controlling SDK behavior.
 * Hosts register features (name + version) to enable conditional card content.
 *
 * ```kotlin
 * val flags = FeatureFlags()
 * flags.register("myFeature", "1.0")
 * config.featureFlags = flags
 * ```
 */
class FeatureFlags {
    private val features = mutableMapOf<String, String>()

    /** Register a feature with a version string. */
    fun register(name: String, version: String) {
        features[name] = version
    }

    /** Unregister a feature. */
    fun unregister(name: String) {
        features.remove(name)
    }

    /** Check if a feature is registered. */
    fun isRegistered(name: String): Boolean = features.containsKey(name)

    /** Get the version of a registered feature, if any. */
    fun version(name: String): String? = features[name]

    /** All registered features. */
    val allFeatures: Map<String, String> get() = features.toMap()

    /**
     * Check if host capabilities meet the given requirements.
     * Per Adaptive Cards spec: `"*"` means any version of the feature must be registered;
     * a specific version string means exact match. ALL requirements must be met.
     * Returns `true` if [requires] is null or empty.
     */
    fun meetsRequirements(requires: Map<String, String>?): Boolean {
        if (requires.isNullOrEmpty()) return true
        for ((name, requiredVersion) in requires) {
            val registeredVersion = features[name] ?: return false
            if (requiredVersion != "*" && registeredVersion != requiredVersion) return false
        }
        return true
    }
}

File: ios/Sources/ACCore/FeatureFlags.swift
// Copyright (c) Microsoft Corporation. All rights reserved.
// Author: Vikrant Singh (github.com/VikrantSingh01)
// Licensed under the MIT License.

import Foundation

/// Feature flags for controlling SDK behavior.
/// Hosts register features (name + version) to enable conditional card content.
///
/// ```swift
/// var flags = FeatureFlags()
/// flags.register(name: "myFeature", version: "1.0")
/// config.featureFlags = flags
/// ```
public struct FeatureFlags: Sendable {
    private var features: [String: String] = [:]

    public init() {}

    /// Register a feature with a version string.
    public mutating func register(name: String, version: String) {
        features[name] = version
    }

    /// Unregister a feature.
    public mutating func unregister(name: String) {
        features.removeValue(forKey: name)
    }

    /// Check if a feature is registered.
    public func isRegistered(name: String) -> Bool {
        features[name] != nil
    }

    /// Get the version of a registered feature, if any.
    public func version(for name: String) -> String? {
        features[name]
    }

    /// All registered features.
    public var allFeatures: [String: String] {
        features
    }

    /// Check if host capabilities meet the given requirements.
    /// Per Adaptive Cards spec: `"*"` means any version of the feature must be registered;
    /// a specific version string means exact match. ALL requirements must be met.
    /// Returns `true` if `requires` is nil or empty.
    public func meetsRequirements(_ requires: [String: String]?) -> Bool {
        guard let requires = requires, !requires.isEmpty else { return true }
        for (name, requiredVersion) in requires {
            guard let registeredVersion = features[name] else { return false }
            if requiredVersion != "*" && registeredVersion != requiredVersion {
                return false
            }
        }
        return true
    }
}
